pkg/rdt/rdt.go

```diff
@@ -0,0 +1,248 @@
+/*
+Copyright 2019 Intel Corporation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rdt
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/ghodss/yaml"
+
+	"github.com/intel/cri-resource-manager/pkg/cri/resource-manager/cache"
+	logger "github.com/intel/cri-resource-manager/pkg/log"
+	"github.com/intel/cri-resource-manager/pkg/utils"
+)
+
+const resctrlGroupPrefix = "cri-resmgr."
+
+// RdtControl is the interface managing Intel RDT resources
+type RdtControl interface {
+	// GetClasses returns the names of RDT classes (or resctrl control groups)
+	// available
+	GetClasses() []string
+
+	//
+	SetClassPID(string, string) error
+
+	SetClassContainer(string, cache.Container) error
+
+	// SetConfig re-configures RDT resources
+	SetConfig(string) error
+}
+
+type rdtControl struct {
+	logger.Logger
+
+	info rdtInfo
+	conf RdtConfig
+}
+
+type RdtConfig struct {
+	ResctrlGroups map[string]ResctrlGroupConfig `json:"resctrlGroups"`
+}
+
+func NewRdtControl(resctrlpath string, config string) (RdtControl, error) {
+	var err error
+	r := &rdtControl{Logger: logger.NewLogger("rdt")}
+
+	// Get info from the resctrl filesystem
+	r.info, err = getRdtInfo(resctrlpath)
+	if err != nil {
+		return nil, err
+	}
+	RdtCacheBitmaskNumBits = uint64(r.info.l3.cbmMask.lsZero())
+
+	// Configure resctrl
+	r.conf, err = parseConfData([]byte(config))
+	if err != nil {
+		return nil, err
+	}
+	if err := r.configureResctrl(r.conf); err != nil {
+		return nil, rdtError("configuration failed: %v", err)
+	}
+
+	return r, nil
+}
+
+func (r *rdtControl) GetClasses() []string {
+	ret := make([]string, len(r.conf.ResctrlGroups))
+
+	i := 0
+	for k, _ := range r.conf.ResctrlGroups {
+		ret[i] = k
+		i++
+	}
+	sort.Strings(ret)
+	return ret
+}
+
+func (r *rdtControl) SetClassContainer(class string, container cache.Container) error {
```

Comment:
Again a bit nitpicking, but my gut feeling is that it would be better to keep the pkg/rdt module generic/resmgr-agnostic. AFAICT, this is the only function which pulls in a dependency on `cache` which would be nice to avoid. It can be easily done by moving this function out of here altogether only keeping the single-pid version (which I would rename the other way around to the IMO more intuitive/natural `SetProcessClass`) , or IMO preferably by leaving here a `SetCgroupClass` which reads all pids from the given cgroups task file an applies the class to them.

Reply:
I agree, better to keep rdt generic. not sure even if cgroup name will be good idea.
IMO, rdt should be able to set/get list of pids per class. And code outside should verify that container list of pids correctly intersects with rdt list.

Reply:
Then the easiest would be tp just change SetClassPID to
```
func (r *rdtControl) SetProcessClass(class string, pids ...int) error {
   for _, pid := range pids {
...
```

Reply:
I wonder, can we utilize some methods from libcontainer? this should be more appropriate than trying to do it all on our side.

Reply:
Dropped dependency on `cache` in the new version of the PR. Renamed `SetClassPID()` to `SetProcessClass` as @klihub suggested (and added support for multiple PIDs).

Haven't taken a look at libcontainer yet